test/spinxsk/spinxsk.c

```diff
@@ -1992,6 +2025,308 @@ ProcessPkts(
+    )
+{
+    //
+    // Split the frame into up to 3 MDLs.
+    //
+    ASSERT_FRE(*NumBuffers > 0);
+    *NumBuffers = (RandUlong() % min(*NumBuffers, 3)) + 1;
+
+    if (*NumBuffers == 1) {
+        Buffers[0].DataOffset = Backfill;
+        Buffers[0].DataLength = FrameLength;
+        Buffers[0].BufferLength = FrameBufferSize;
+        Buffers[0].VirtualAddress = Frame;
+    } else if (*NumBuffers == 2) {
+        const UINT32 bufferSplitOffset = RandUlong() % (FrameLength + 1);
+
+        Buffers[0].DataOffset = Backfill;
+        Buffers[0].DataLength = bufferSplitOffset;
+        Buffers[0].BufferLength = Backfill + bufferSplitOffset;
+        Buffers[0].VirtualAddress = Frame;
+
+        Buffers[1].DataLength = FrameLength - bufferSplitOffset;
+        Buffers[1].BufferLength = FrameBufferSize - Buffers[0].BufferLength;
+        Buffers[1].VirtualAddress = Buffers[0].VirtualAddress + Buffers[0].BufferLength;;
+    } else { // *NumBuffers == 3
+        ASSERT_FRE(*NumBuffers == 3);
+        const UINT32 bufferSplitOffset = RandUlong() % (FrameLength + 1);
+        const UINT32 bufferSplitOffset2 =
+            bufferSplitOffset + RandUlong() % (FrameLength - bufferSplitOffset + 1);
+
+        Buffers[0].DataOffset = Backfill;
+        Buffers[0].DataLength = bufferSplitOffset;
+        Buffers[0].BufferLength = Backfill + bufferSplitOffset;
+        Buffers[0].VirtualAddress = Frame;
+
+        Buffers[1].DataLength = bufferSplitOffset2 - bufferSplitOffset;
+        Buffers[1].BufferLength = bufferSplitOffset2 - bufferSplitOffset;
+        Buffers[1].VirtualAddress = Buffers[0].VirtualAddress + Buffers[0].BufferLength;
+
+        Buffers[2].DataLength = FrameLength - bufferSplitOffset2;
+        Buffers[2].BufferLength = FrameBufferSize - Backfill - bufferSplitOffset2;
+        Buffers[2].VirtualAddress = Buffers[1].VirtualAddress + Buffers[1].BufferLength;
+    }
+}
+
+//
+// Build and inject frames in the receive path through the FNMP driver.
+//
+VOID
+InjectFnmpRxFrames(
+    _In_ FNMP_HANDLE Fnmp,
+    _In_ const UINT32 NumFrames
+    )
+{
+    UINT32 payloadBufferSize = 0;
+    DATA_FLUSH_OPTIONS flushOptions = {0};
+    ULONG frameBufferSize = 0;
+    UCHAR* payload = NULL;
+    UCHAR* frame = NULL;
+    HRESULT status = S_OK;
+
+    //
+    // Use a very large payload buffer from time to time
+    // but keep it smaller most of the time to iterate faster:
+    // most of the interesting logic has to do with the headers.
+    //
+    if (RandUlong() % 10 == 0)
+    {
+        payloadBufferSize = 64'000;
+    } else {
+        payloadBufferSize = 500;
+    }
+    frameBufferSize = FNMP_FRAME_MAX_BACKFILL + TCP_HEADER_STORAGE + payloadBufferSize;
+
+    payload = calloc(payloadBufferSize, sizeof(UCHAR));
+    frame = calloc(frameBufferSize, sizeof(UCHAR));
+
+    if (frame == NULL || payload == NULL) {
+        TraceWarn("Allocation failure, frame: %p, payload: %p", frame, payload);
+        free(frame);
```

Comment:
nit (sorry for catching this late): please `goto cleanup` rather than duplicating the `free` calls here.